cmd/rpcdaemon/commands/eth_api.go

```diff
@@ -269,6 +274,47 @@ func (api *BaseAPI) headerByRPCNumber(number rpc.BlockNumber, tx kv.Tx) (*types.
 	return api._blockReader.Header(context.Background(), tx, h, n)
 }
 
+// checks the pruning state to see if we would hold information about this
+// block in state history or not.  Some strange issues arise getting account
+// history for blocks that have been pruned away giving nonce too low errors
+// etc. as red herrings
+func (api *BaseAPI) checkPruneHistory(tx kv.Tx, block uint64) error {
+	p, err := api.loadPruneHistory(tx)
+	if err != nil {
+		return err
+	}
+	if p.History.Enabled() {
+		latest, err := api.blockByRPCNumber(rpc.LatestBlockNumber, tx)
+		if err != nil {
+			return err
+		}
+		prunedTo := p.History.PruneTo(latest.Number().Uint64())
+		if block < prunedTo {
+			return fmt.Errorf("history has been pruned for this block")
+		}
+	}
+
+	return nil
+}
+
+func (api *BaseAPI) loadPruneHistory(tx kv.Tx) (*prune.Mode, error) {
+	if api._pruneMode != nil {
```

Comment:
write inside mutex, but read without mutex - it's race condition.
see `historyV3()` as an example

Reply:
Good shout, sorted now